### PR TITLE
[Contracts] move Translatable to symfony/translation-contracts

### LINKED ISSUE #38319: [RFC] Move Translatable to contracts
Currently the new `Translatable` value object and its corresponding `t()` factory, live in the `Translator` component.

Given its general purpose usability for any translator implementation, i suggest to move both symbols to the `Translation` contract.

It seems more reasonable to leak contracts into our domain layer, rather than an infrastructural component.

Besides, this provides a defacto standard for translation extractors to act upon as well.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #38319
| License       | MIT
| Doc PR        | -

As discussed in the RFC:

- `Translatable` is moved to symfony/contracts;
- `Translatable` is not final anymore, it makes no sense for a value-object, especially in contracts;
- the `t()` function is kept in the component - it doesn't fit in the contracts IMHO;
- `Translatable::trans()` is not static anymore;
- the domain is nullable instead of defaulting to "messages";
- the `t()` function moved in the `Symfony\Component\Translation` namespace (for reference, the `s()` function from String is also namespaced.)

### CODE REVIEW

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Bridge/Twig/Extension/TranslationExtension.php [hunk 1]
```diff
@@ -125,7 +125,7 @@ public function trans($message, array $arguments = [], string $domain = null, st
         return $this->getTranslator()->trans($message, $arguments, $domain, $locale);
     }
 
-    public function createTranslatable(string $message, array $parameters = [], string $domain = 'messages'): Translatable
+    public function createTranslatable(string $message, array $parameters = [], string $domain = null): Translatable
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/TranslationExtension.php:128]: in `trans()`, shouldn't we support using the `$domain` argument explicitly when the `Translatable` object does not have a domain ? Currently, we always override the domain argument on line 113.

**Comment by nicolas-grekas** [comment]
FYI, I just moved the `t()` function in the `Symfony\Component\Translation` namespace (for reference, the `s()` function from String is also namespaced.)

File: src/Symfony/Component/Translation/Resources/functions/translatable.php [hunk 2]
```diff
@@ -1,22 +0,0 @@
...
- *
- * (c) Fabien Potencier <fabien@symfony.com>
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
-
-use Symfony\Component\Translation\Translatable;
-
-if (!function_exists('t')) {
```
Comment (stof) [inline src/Symfony/Component/Translation/Resources/functions/translatable.php:14]: removing this `function_exists` can cause conflicts with other projects (this function is not namespaced)

Comment (stof) [reply src/Symfony/Component/Translation/Resources/functions/translatable.php:14]: Forget about that one. You added a namespace in parallel of my review...

File: src/Symfony/Bridge/Twig/Extension/TranslationExtension.php [hunk 3]
```diff
@@ -104,17 +104,17 @@ public function getTranslationNodeVisitor(): TranslationNodeVisitor
...
     /**
-     * @param ?string|Translatable $message The message id (may also be an object that can be cast to string)
+     * @param string|\Stringable|Translatable|null $message
      */
     public function trans($message, array $arguments = [], string $domain = null, string $locale = null, int $count = null): string
     {
         if ($message instanceof Translatable) {
             $arguments += $message->getParameters();
-            $domain = $message->getDomain();
+            $domain = $domain ?? $message->getDomain();
```
Comment (ro0NL) [inline src/Symfony/Bridge/Twig/Extension/TranslationExtension.php:113]: we reasoned the domain is bound to a translatable, and is not intended to be overridden dynamically.

So personally i'd prefer `$message->getDomain ?? $domain`, but this needs to be re-evaluted perhaps.

File: src/Symfony/Bridge/Twig/Extension/TranslationExtension.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/TranslationExtension.php:128]: L115 updated to `$domain = $domain ?? $message->getDomain();`

File: src/Symfony/Bridge/Twig/Extension/TranslationExtension.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/TranslationExtension.php:113]: fine to me, I've swapped these

**Review by stof — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
See alternative proposal in #38330.
It's one or the other :)

**Comment by stof** [comment]
@nicolas-grekas maybe a separate PR could be created (and merged) with all the common changes (nullable domain, namespaced `t()`, non static `trans()` method, etc...), leaving the 2 alternative PRs with only the changes related to their own approach (this depends how long the discussion about alternatives will take).

On a side note, merging any of these PRs will require updating the blog post as it changes the API.

**Comment by nicolas-grekas** [comment]
Closing in favor of #38330 since that's where the consensus seems to go.

**Pull request closed without merging**